internal/report/report.go

```diff
@@ -828,6 +831,55 @@ func printText(w io.Writer, rpt *Report) error {
 	return nil
 }
 
+// printFolded prints a profile in Brendan Gregg's Folded Stacks format.
+func printFolded(w io.Writer, rpt *Report) error {
+	prof := rpt.prof
+	o := rpt.options
+
+	_, locations := graph.CreateNodes(prof, &graph.Options{})
+	for _, sample := range prof.Sample {
+		var stack []*graph.NodeInfo
+		for _, loc := range sample.Location {
+			nodes := locations[loc.ID]
+			for _, n := range nodes {
+				stack = append(stack, &n.Info)
+			}
+		}
+
+		if len(stack) == 0 {
+			continue
+		}
+
+		var d, v int64
+		v = o.SampleValue(sample.Value)
+		if o.SampleMeanDivisor != nil {
+			d = o.SampleMeanDivisor(sample.Value)
+		}
+		if d != 0 {
+			v = v / d
+		}
+		// Print call stack.
+		for i := range stack {
+			// Folded stack convention: start with root frame, end
+			// with leaves.
+			s := stack[len(stack)-i-1]
+			if i > 0 {
+				fmt.Fprint(w, ";")
+			}
+			// TODO: should we print more than just s.Name?
```

Comment:
Yes, I think this needs more discussion. There is an implicit interplay with the granularity. See below. I would expect that we utilize that - i.e. use PrintableName() or NameComponents() directly defining how they should join. This is in particular important because graph.CreateNodes() respects the granularity and so if we don't encode all components in the string, then something like "-folded -filefunctions" may produce duplicate elements.

```
$ pprof -top -nodecount=5 -symbolize=none internal/report/testdata/sample.cpu
File: sample.bin
Type: cpu
Time: Sep 29, 2017 at 1:57pm (PDT)
Duration: 2s, Total samples = 1760ms (87.91%)
Showing nodes accounting for 1640ms, 93.18% of 1760ms total
Showing top 5 nodes out of 14
      flat  flat%   sum%        cum   cum%
    1280ms 72.73% 72.73%     1370ms 77.84%  runtime.mapiternext
     150ms  8.52% 81.25%     1760ms   100%  main.busyLoop
     110ms  6.25% 87.50%      110ms  6.25%  math.Abs (inline)
      60ms  3.41% 90.91%       70ms  3.98%  runtime.evacuate
      40ms  2.27% 93.18%       40ms  2.27%  runtime.add (inline)
$ pprof -top -nodecount=5 -symbolize=none -filefunctions internal/report/testdata/sample.cpu
File: sample.bin
Type: cpu
Time: Sep 29, 2017 at 1:57pm (PDT)
Duration: 2s, Total samples = 1760ms (87.91%)
Showing nodes accounting for 1640ms, 93.18% of 1760ms total
Showing top 5 nodes out of 14
      flat  flat%   sum%        cum   cum%
    1280ms 72.73% 72.73%     1370ms 77.84%  runtime.mapiternext /usr/lib/google-golang/src/runtime/hashmap.go
     150ms  8.52% 81.25%     1760ms   100%  main.busyLoop internal/report/testdata/sample/sample.go
     110ms  6.25% 87.50%      110ms  6.25%  math.Abs /usr/lib/google-golang/src/math/abs.go (inline)
      60ms  3.41% 90.91%       70ms  3.98%  runtime.evacuate /usr/lib/google-golang/src/runtime/hashmap.go
      40ms  2.27% 93.18%       40ms  2.27%  runtime.add /usr/lib/google-golang/src/runtime/stubs.go (inline)
$ pprof -top -nodecount=5 -symbolize=none -files internal/report/testdata/sample.cpu
File: sample.bin
Type: cpu
Time: Sep 29, 2017 at 1:57pm (PDT)
Duration: 2s, Total samples = 1760ms (87.91%)
Showing nodes accounting for 1740ms, 98.86% of 1760ms total
Showing top 5 nodes out of 8
      flat  flat%   sum%        cum   cum%
    1400ms 79.55% 79.55%     1450ms 82.39%  /usr/lib/google-golang/src/runtime/hashmap.go
     150ms  8.52% 88.07%     1760ms   100%  internal/report/testdata/sample/sample.go
     110ms  6.25% 94.32%      110ms  6.25%  /usr/lib/google-golang/src/math/abs.go (inline)
      40ms  2.27% 96.59%      130ms  7.39%  /usr/lib/google-golang/src/runtime/hashmap_fast.go
      40ms  2.27% 98.86%       40ms  2.27%  /usr/lib/google-golang/src/runtime/stubs.go (inline)
$ pprof -top -nodecount=5 -symbolize=none -addresses internal/report/testdata/sample.cpu
File: sample.bin
Type: cpu
Time: Sep 29, 2017 at 1:57pm (PDT)
Duration: 2s, Total samples = 1760ms (87.91%)
Showing nodes accounting for 350ms, 19.89% of 1760ms total
Showing top 5 nodes out of 96
      flat  flat%   sum%        cum   cum%
     100ms  5.68%  5.68%      100ms  5.68%  00000000004b4463 math.Abs /usr/lib/google-golang/src/math/abs.go:16
      70ms  3.98%  9.66%       70ms  3.98%  00000000004095d0 runtime.mapiternext /usr/lib/google-golang/src/runtime/hashmap.go:823
      70ms  3.98% 13.64%       70ms  3.98%  00000000004096a2 runtime.mapiternext /usr/lib/google-golang/src/runtime/hashmap.go:856
      60ms  3.41% 17.05%       60ms  3.41%  000000000040957a runtime.mapiternext /usr/lib/google-golang/src/runtime/hashmap.go:819
      50ms  2.84% 19.89%       50ms  2.84%  00000000004095b6 runtime.mapiternext /usr/lib/google-golang/src/runtime/hashmap.go:822
```